### ISSUE TITLE
Panic while running terraform plan after state move

### ISSUE THREAD

**Issue by ryanhartkopf** [hashicorp/terraform#19189]
While working on moving a resource into a module, I made a typo in my `terraform state mv` command. The next time I ran `terraform plan` I got a SIGSEGV error. I have been unable to revert the changes to the state file using built-in commands.

This issue happens when the resource types in the `terraform state mv` command do not match.

### Terraform Version

```
Terraform v0.11.10
+ provider.aws v1.39.0
```

### Terraform Configuration Files

```hcl
provider "aws" {
  region  = "us-east-1"
  version = "~> 1.39.0"
}

resource "aws_s3_bucket" "foo" {
  bucket = "foo-v45qnpbavw548n"
}
```

### Crash Output

https://gist.github.com/ryanhartkopf/97c96750a0ea812b0b945d0c8dbde665
### Expected Behavior

Successful terraform plan.
### Actual Behavior

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x17668c8]
```
### Steps to Reproduce

```
terraform init
terraform apply
terraform state mv aws_s3_bucket.foo aws_s3_bucket_policy.foo
terraform plan
```

I'm unable to fix by running `terraform state mv` or `terraform state rm`. The bad resource remains in both cases.

### References

#17659 seems like it may be close.

**Comment by ryanhartkopf** [comment]
Here is a diff of before and after I ran the command `terraform state move aws_iam_policy.copy_ami_from_prod module.iam_jenkins.aws_iam_role_policy_attachment.copy_ami_from_prod`.

You can see that the resource type in the resource name and the "type" attribute do not match. I checked my most recent state file and the resource is still present with the incorrect type, and I can't remove it or rename it with any of the `terraform state` commands. I'm not sure if this is causing the crash but it's the only lead I have.

**Comment by ryanhartkopf** [comment]
I was able to replicate this issue very easily with a single S3 bucket.

```
provider "aws" {
  region  = "us-east-1"
  version = "~> 1.39.0"
}

resource "aws_s3_bucket" "foo" {
  bucket = "foo-v45qnpbavw548n"
}
```

```
tf-state-test$ terraform init       
Initializing provider plugins...                                     
- Checking for available provider plugins on https://releases.hashicorp.com...
- Downloading plugin for provider "aws" (1.39.0)...
                                                  
Terraform has been successfully initialized!  
                                                                        
You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.                                 
                                            
If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
tf-state-test$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.
                                                             
                                                                  
------------------------------------------------------------------------
                                                                       
An execution plan has been generated and is shown below.                                 
Resource actions are indicated with the following symbols:        
  + create                                           
                                                                        
Terraform will perform the following actions:

  + aws_s3_bucket.foo
      id:                          <computed>
      acceleration_status:         <computed>
      acl:                         "private"
      arn:                         <computed>
      bucket:                      "foo-v45qnpbavw548n"
      bucket_domain_name:          <computed>
      bucket_regional_domain_name: <computed>
      force_destroy:               "false"
      hosted_zone_id:              <computed>
      region:                      <computed>
      request_payer:               <computed>
      versioning.#:                <computed>
      website_domain:              <computed>
      website_endpoint:            <computed>


Plan: 1 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

This plan was saved to: planned

To perform exactly these actions, run the following command to apply:
    terraform apply "planned"

tf-state-test$ terraform apply
aws_s3_bucket.foo: Creating...
  acceleration_status:         "" => "<computed>"
  acl:                         "" => "private"
  arn:                         "" => "<computed>"
  bucket:                      "" => "foo-v45qnpbavw548n"
  bucket_domain_name:          "" => "<computed>"
  bucket_regional_domain_name: "" => "<computed>"
  force_destroy:               "" => "false"
  hosted_zone_id:              "" => "<computed>"
  region:                      "" => "<computed>"
  request_payer:               "" => "<computed>"
  versioning.#:                "" => "<computed>"
  website_domain:              "" => "<computed>"
  website_endpoint:            "" => "<computed>"
aws_s3_bucket.foo: Creation complete after 3s (ID: foo-v45qnpbavw548n)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
tf-state-test$ terraform state move aws_s3_bucket.foo aws_s3_bucket_policy.foo
Moved aws_s3_bucket.foo to aws_s3_bucket_policy.foo
tf-state-test$ terraform state move aws_s3_bucket_policy.foo aws_s3_bucket.foo
Item to move doesn't exist: aws_s3_bucket_policy.foo
tf-state-test$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x17668c8]

goroutine 178 [running]:
github.com/hashicorp/terraform/terraform.(*NodeRefreshableManagedResourceInstance).evalTreeManagedResourceNoState(0xc0001e42a8, 0xa434438, 0xc0001de4e0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/node_resource_refresh.go:238 +0x328
github.com/hashicorp/terraform/terraform.(*NodeRefreshableManagedResourceInstance).EvalTree(0xc0001e42a8, 0x3acf1c0, 0xc0001e42a8)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/node_resource_refresh.go:98 +0x6e
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x3acf1c0, 0xc0001e42a8, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph.go:113 +0x9a2
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000222460, 0x3acf1c0, 0xc0001e42a8, 0xc000466d80)
        /opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:387 +0x367
created by github.com/hashicorp/terraform/dag.(*Walker).Update
        /opt/gopath/src/github.com/hashicorp/terraform/dag/walk.go:310 +0x986



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

**Comment by ryanhartkopf** [comment]
I see that this will be fixed in v0.12: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/terraform-tool/B6pN4izPS4g

> I'm currently updating (rewriting) the `state` commands to make them work in v0.12 and in general we try to keep the behavior of the commands inline with how they worked before v0.12. But we did make a few small improvements to hopefully fix a few small shortcomings. One of those fixes is that when you try to use `state mv` to move a resource, we now first check if the source and destination resources are of the same type.

There are _some_ resources that it would be nice to be able to rename the type and retain the same attributes, such as renaming `aws_alb` to `aws_lb` as shown in #16633.
